apollo-federation/src/sources/connect/models/validation.rs

```diff
@@ -49,3 +39,225 @@
+}
+
+/// A `@source` directive along with any errors related to it.
+struct SourceDirective {
+    name: SourceName,
+    errors: Vec<ValidationError>,
+}
+
+/// The `name` argument of a `@source` directive.
+#[derive(Clone, Debug)]
+pub enum SourceName {
+    /// A perfectly reasonable source name.
+    Valid(String),
+    /// Contains invalid characters, so it will have to be renamed. This means certain checks
+    /// (like uniqueness) should be skipped. However, we have _a_ name, so _other_ checks on the
+    /// `@source` directive can continue.
+    Invalid(String),
+}
+
+impl From<SourceName> for String {
+    fn from(source_name: SourceName) -> String {
+        match source_name {
+            SourceName::Valid(name) | SourceName::Invalid(name) => name,
+        }
+    }
+}
+
+impl TryFrom<&Component<Directive>> for SourceName {
+    type Error = ValidationError;
+
+    fn try_from(directive: &Component<Directive>) -> Result<Self, Self::Error> {
+        let str_value = directive
+            .arguments
+            .iter()
+            .find(|arg| arg.name == "name")
+            .ok_or(ValidationError::MissingSourceName)?
+            .value
+            .as_str()
+            .ok_or(ValidationError::SourceNameType)?;
+        if str_value.chars().all(|c| c.is_alphanumeric() || c == '_') {
+            Ok(Self::Valid(str_value.to_string()))
+        } else {
+            Ok(Self::Invalid(str_value.to_string()))
+        }
+    }
+}
+
+impl Display for SourceName {
+    fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
+        match self {
+            Self::Valid(name) | Self::Invalid(name) => write!(f, "{}", name),
+        }
+    }
+}
+
+#[derive(Debug, thiserror::Error)]
+pub enum ValidationError {
+    #[error("baseURL argument for @source \"{source_name}\" was not a valid URL: {inner}")]
+    SourceUrl {
+        inner: url::ParseError,
+        source_name: SourceName,
+    },
+    #[error("baseURL argument for @source \"{source_name}\" be http or https, got {scheme}")]
+    SourceScheme {
+        scheme: String,
+        source_name: SourceName,
+    },
+    #[error("Missing name argument for a @source directive")]
+    MissingSourceName,
+    #[error("name argument for a @source directive must be a string")]
+    SourceNameType,
+    #[error(
+        "Invalid characters in source name {0}. Only alphanumeric and underscores are allowed."
+    )]
+    InvalidSourceName(SourceName),
+    /// Intentionally a String so we remember to only check valid source names
+    #[error("Every @source name must be unique; found duplicate source name {0}")]
+    DuplicateSourceName(String),
+}
+#[cfg(test)]
+mod test_validate_source {
+    use super::*;
+
+    #[test]
+    fn unparseable_url() {
+        let schema = include_str!("test_data/invalid_source_url.graphql");
+        let schema = Schema::parse(schema, "test.graphql").unwrap();
+        let errors = validate(schema);
+        assert_eq!(errors.len(), 1);
```

Comment:
Nit: It might be good to check that the inner values of these tests include the right information vs just checking that they are of the right type.

Reply:
I think I want to enhance these with a snapshot mechanism so it doesn't get too tedious to update the source schema files for tests. I _think_ insta can do inline snapshots, I'll do that as a follow-up.

Reply:
inline snapshots are easy! `assert_snapshot(whatever, @"")`